// hadouken-js-adapter/out/src/api/fin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const system_1 = require("./system/system");
const window_1 = require("./window/window");
const application_1 = require("./application/application");
const interappbus_1 = require("./interappbus/interappbus");
const notification_1 = require("./notification/notification");
const clipboard_1 = require("./clipboard/clipboard");
const external_application_1 = require("./external-application/external-application");
const frame_1 = require("./frame/frame");
const global_hotkey_1 = require("./global-hotkey");
class Fin extends events_1.EventEmitter {
    get me() {
        return this.wire.me;
    }
    constructor(wire) {
        super();
        this.wire = wire;
        this.System = new system_1.default(wire);
        this.Window = new window_1.default(wire);
        this.Application = new application_1.default(wire);
        this.InterApplicationBus = new interappbus_1.default(wire);
        this.Notification = new notification_1.default(wire);
        this.Clipboard = new clipboard_1.default(wire);
        this.ExternalApplication = new external_application_1.default(wire);
        this.Frame = new frame_1.default(wire);
        this.GlobalHotkey = new global_hotkey_1.default(wire);
        //Handle disconnect events
        wire.on('disconnected', () => {
            this.emit('disconnected');
        });
    }
}
exports.default = Fin;
