// hadouken-js-adapter/out/src/api/interappbus/channel/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const channel_1 = require("./channel");
class ChannelProvider extends channel_1.ChannelBase {
    constructor(providerIdentity, send) {
        super(providerIdentity, send);
        this.connectListener = () => undefined;
        this.disconnectListener = () => undefined;
        this.connections = [];
    }
    dispatch(to, action, payload) {
        return this.send(to, action, payload);
    }
    async processConnection(senderId, payload) {
        this.connections.push(senderId);
        return this.connectListener(senderId, payload);
    }
    publish(action, payload) {
        return this.connections.map(to => this.send(to, action, payload));
    }
    onConnection(listener) {
        this.connectListener = listener;
    }
    onDisconnection(listener) {
        this.disconnectListener = listener;
    }
    async destroy() {
        const { channelName } = this.providerIdentity;
        await this.sendRaw('destroy-channel', { channelName });
        const { channelId } = this.providerIdentity;
        this.removeChannel(channelId);
    }
}
exports.ChannelProvider = ChannelProvider;
